File: packages/vela-sdk/src/vela_sdk/fastmcp/session_elicitor.py
"""Session elicitation helpers — param collection and resume-or-new flows."""

from typing import Optional

import structlog
from fastmcp.server.elicitation import AcceptedElicitation

from vela_sdk.engine.types import WorkflowRunState
from vela_sdk.fastmcp.locale import Locale, get_locale
from vela_sdk.fastmcp.response_builder import build_run_options
from vela_sdk.schemas.workflow import WorkflowDefinition

logger = structlog.get_logger()


async def elicit_required_params(
    ctx,
    param_defs: list,
    locale: Optional[Locale] = None,
) -> Optional[dict]:
    """Elicit values for required params. Returns dict or None on cancel."""
    result: dict = {}
    for p_def in param_defs:
        label = p_def.label or p_def.name
        description = f" — {p_def.description}" if p_def.description else ""
        message = f"{label}{description}"
        if p_def.default is not None:
            message += f" [default: {p_def.default}]"

        try:
            elicit_result = await ctx.elicit(message, str)
        except Exception:
            logger.debug("elicitation.not_supported", key=p_def.name)
            return {}

        if isinstance(elicit_result, AcceptedElicitation) and elicit_result.data:
            result[p_def.name] = elicit_result.data
        elif p_def.default is not None:
            result[p_def.name] = p_def.default
        else:
            return None
    return result


async def elicit_session_choice(
    ctx,
    wf_def: WorkflowDefinition,
    active_runs: list[WorkflowRunState],
    locale: Optional[Locale] = None,
) -> tuple[Optional[WorkflowRunState], Optional[str]]:
    """Elicit resume-or-new choice."""
    if locale is None:
        locale = get_locale()

    options = build_run_options(wf_def, active_runs, locale=locale)

    try:
        result = await ctx.elicit(
            locale.session_choice_message.format(wf_name=wf_def.name),
            options,
        )
    except Exception:
        logger.debug("elicitation.not_supported", workflow_id=wf_def.id)
        return None, None

    if not isinstance(result, AcceptedElicitation):
        return None, None

    if result.data == "__new__":
        return None, "__new__"

    chosen_run = next((r for r in active_runs if r.id == result.data), None)
    return chosen_run, None


async def elicit_prompt_session(
    ctx,
    wf_def: WorkflowDefinition,
    active_runs: list[WorkflowRunState],
    locale: Optional[Locale] = None,
) -> tuple[Optional[WorkflowRunState], dict]:
    """Elicit session choice in prompt handler."""
    if active_runs:
        chosen_run, choice = await elicit_session_choice(ctx, wf_def, active_runs, locale=locale)
        if chosen_run:
            return chosen_run, {}
        if choice != "__new__":
            return None, {}

    required_params = [p for p in wf_def.params if p.required]
    new_params = await elicit_required_params(ctx, required_params, locale=locale)
    return None, new_params or {}


async def elicit_missing_params(
    ctx,
    wf_def: WorkflowDefinition,
    missing_params: list,
    active_runs: list[WorkflowRunState],
    existing_params: dict,
    locale: Optional[Locale] = None,
) -> Optional[dict]:
    """Elicit missing required/identity params via resume-or-new flow."""
    if active_runs:
        chosen_run, choice = await elicit_session_choice(ctx, wf_def, active_runs, locale=locale)
        if chosen_run:
            run_params = chosen_run.params
            return {
                p_def.name: run_params[p_def.name]
                for p_def in missing_params
                if p_def.name in run_params
            }
        if choice is None:
            return None if chosen_run is None else {}

    still_missing = [
        p for p in missing_params
        if p.name not in existing_params
    ]
    return await elicit_required_params(ctx, still_missing, locale=locale)

File: packages/vela-sdk-ts/src/mcp/session-elicitor.ts
/**
 * Session elicitation helpers — param collection and resume-or-new flows.
 *
 * Provides functions for eliciting required workflow params, choosing
 * between active sessions, and handling the full missing-params flow.
 */

import type { WorkflowRunState } from "../engine/types.js";
import type { Locale } from "../locale/locale.js";
import { getLocale } from "../locale/locale.js";
import type {
  ParamDefinition,
  WorkflowDefinition,
} from "../schemas/workflow.js";
import type { McpContext } from "./mcp-server.js";
import { buildRunOptions } from "./response-builder.js";

// ---------------------------------------------------------------------------
// elicitRequiredParams
// ---------------------------------------------------------------------------

/**
 * Elicit values for required params.
 *
 * Returns a dict of param values, or `null` if the user cancels.
 * Returns `{}` if elicitation is not supported.
 */
export async function elicitRequiredParams(
  ctx: McpContext,
  paramDefs: ParamDefinition[],
  locale?: Locale,
): Promise<Record<string, unknown> | null> {
  const result: Record<string, unknown> = {};

  for (const pDef of paramDefs) {
    const label = pDef.label ?? pDef.name;
    const description = pDef.description ? ` \u2014 ${pDef.description}` : "";
    let message = `${label}${description}`;
    if (pDef.default !== undefined && pDef.default !== null) {
      message += ` [default: ${pDef.default}]`;
    }

    const schema: Record<string, unknown> = {
      type: "object",
      properties: {
        value: { type: "string", title: label },
      },
      required: ["value"],
    };

    let elicitResult;
    try {
      elicitResult = await ctx.elicit(message, schema);
    } catch {
      ctx.log.debug("elicitation.not_supported", { key: pDef.name });
      return {};
    }

    if (
      elicitResult.action === "accept" &&
      elicitResult.content &&
      "value" in elicitResult.content &&
      elicitResult.content["value"]
    ) {
      result[pDef.name] = elicitResult.content["value"];
    } else if (pDef.default !== undefined && pDef.default !== null) {
      result[pDef.name] = pDef.default;
    } else {
      return null;
    }
  }

  return result;
}

// ---------------------------------------------------------------------------
// elicitSessionChoice
// ---------------------------------------------------------------------------

/**
 * Elicit a resume-or-new choice from the user.
 *
 * Returns `[chosenRun, null]` if an existing run was selected,
 * `[null, "__new__"]` if the user chose to start new, or
 * `[null, null]` if elicitation was cancelled or not supported.
 */
export async function elicitSessionChoice(
  ctx: McpContext,
  wfDef: WorkflowDefinition,
  activeRuns: WorkflowRunState[],
  locale?: Locale,
): Promise<[WorkflowRunState | null, string | null]> {
  const loc = locale ?? getLocale();

  const options = buildRunOptions(wfDef, activeRuns, loc);

  const schema: Record<string, unknown> = {
    type: "object",
    properties: {
      value: {
        type: "string",
        title: loc.sessionChoiceMessage.replace("{wf_name}", wfDef.name),
        enum: Object.keys(options),
        "x-enum-descriptions": Object.fromEntries(
          Object.entries(options).map(([k, v]) => [k, v.title]),
        ),
      },
    },
    required: ["value"],
  };

  let result;
  try {
    result = await ctx.elicit(
      loc.sessionChoiceMessage.replace("{wf_name}", wfDef.name),
      schema,
    );
  } catch {
    ctx.log.debug("elicitation.not_supported", {
      workflow_id: wfDef.id,
    });
    return [null, null];
  }

  if (result.action !== "accept" || !result.content) {
    return [null, null];
  }

  const chosen = result.content["value"] as string | undefined;
  if (chosen === "__new__") {
    return [null, "__new__"];
  }

  const chosenRun = activeRuns.find((r) => r.id === chosen) ?? null;
  return [chosenRun, null];
}

// ---------------------------------------------------------------------------
// elicitPromptSession
// ---------------------------------------------------------------------------

/**
 * Elicit session choice in prompt handler.
 *
 * Returns `[existingRun, {}]` if resuming, or `[null, params]` for new.
 */
export async function elicitPromptSession(
  ctx: McpContext,
  wfDef: WorkflowDefinition,
  activeRuns: WorkflowRunState[],
  locale?: Locale,
): Promise<[WorkflowRunState | null, Record<string, unknown>]> {
  if (activeRuns.length > 0) {
    const [chosenRun, choice] = await elicitSessionChoice(
      ctx,
      wfDef,
      activeRuns,
      locale,
    );
    if (chosenRun) {
      return [chosenRun, {}];
    }
    if (choice !== "__new__") {
      return [null, {}];
    }
  }

  const requiredParams = wfDef.params.filter((p) => p.required);
  const newParams = await elicitRequiredParams(ctx, requiredParams, locale);
  return [null, newParams ?? {}];
}

// ---------------------------------------------------------------------------
// elicitMissingParams
// ---------------------------------------------------------------------------

/**
 * Elicit missing required/identity params via resume-or-new flow.
 *
 * Returns a dict of param values, or `null` if cancelled.
 */
export async function elicitMissingParams(
  ctx: McpContext,
  wfDef: WorkflowDefinition,
  missingParams: ParamDefinition[],
  activeRuns: WorkflowRunState[],
  existingParams: Record<string, unknown>,
  locale?: Locale,
): Promise<Record<string, unknown> | null> {
  if (activeRuns.length > 0) {
    const [chosenRun, choice] = await elicitSessionChoice(
      ctx,
      wfDef,
      activeRuns,
      locale,
    );
    if (chosenRun) {
      const runParams = chosenRun.params;
      const result: Record<string, unknown> = {};
      for (const pDef of missingParams) {
        if (pDef.name in runParams) {
          result[pDef.name] = runParams[pDef.name];
        }
      }
      return result;
    }
    if (choice === null) {
      return null;
    }
  }

  const stillMissing = missingParams.filter(
    (p) => !(p.name in existingParams),
  );
  return elicitRequiredParams(ctx, stillMissing, locale);
}
